<a href="https://colab.research.google.com/github/davensethi/DS-Unit-1-Sprint-1-Data-Wrangling-and-Storytelling/blob/master/DS-Unit-1-Sprint-2-Statistics/module2/LS_DS_121_Statistics_Probability_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>
<br></br>
<br></br>

## *Data Science Unit 1 Sprint 2 Assignment 1*

# Apply the t-test to real data

## Practice 1-Sample T-tests

One Sample t-tests determine whether or not a sample mean is statistically different from some known (or hypothesized) population mean. 

### 1) Load the Data
- Use the [automobile dataset](https://archive.ics.uci.edu/ml/datasets/Automobile)
- Fix the column headers
- Make sure NaNs are used to indicate missing values

Feel free to add code cells and text cells as needed throughout the assignment.

In [147]:
import pandas as pd
import numpy as np



In [151]:
from google.colab import files
uploaded = files.upload()

Saving imports-85.data to imports-85 (1).data


In [152]:

column_headers =['symboling', 'normalized-losses','make', 'fuel-type','aspiration',
                 'num-of-doors','body-style', 'drive-wheels','engine-location','wheel-base',
                 'length','width','height','curb-weight', 'engine-type','num-of-cylinders',
                 'engine-size','fuel-system' ,'bore' ,'stroke' ,'compression-ratio:' ,'horsepower' ,
                 'peak-rpm' ,'city-mpg','highway-mpg' ,'price']

df = pd.read_csv('imports-85.data', names= column_headers)
df.head(10)

symboling normalized-losses         make  ... city-mpg highway-mpg  price
0          3                 ?  alfa-romero  ...       21          27  13495
1          3                 ?  alfa-romero  ...       21          27  16500
2          1                 ?  alfa-romero  ...       19          26  16500
3          2               164         audi  ...       24          30  13950
4          2               164         audi  ...       18          22  17450
5          2                 ?         audi  ...       19          25  15250
6          1               158         audi  ...       19          25  17710
7          1                 ?         audi  ...       19          25  18920
8          1               158         audi  ...       17          20  23875
9          0                 ?         audi  ...       16          22      ?

[10 rows x 26 columns]

In [153]:
df = df.replace({'?':np.NaN})

df.head(10)

symboling normalized-losses         make  ... city-mpg highway-mpg  price
0          3               NaN  alfa-romero  ...       21          27  13495
1          3               NaN  alfa-romero  ...       21          27  16500
2          1               NaN  alfa-romero  ...       19          26  16500
3          2               164         audi  ...       24          30  13950
4          2               164         audi  ...       18          22  17450
5          2               NaN         audi  ...       19          25  15250
6          1               158         audi  ...       19          25  17710
7          1               NaN         audi  ...       19          25  18920
8          1               158         audi  ...       17          20  23875
9          0               NaN         audi  ...       16          22    NaN

[10 rows x 26 columns]

### 2) Pretend that this dataset represents the cars at a used car lot in your local town. 

- Use df.sample() to pick a random sample of 10 cars. Note that because this sample is random we are going to set the `random_state` so that all of us in the class get the same random sample. Please set your random state to `30` when using `df.sample()`

Is your sample reflective of the population value in regards to highway-mpg? Find the mean for `highway-mpg` for the entire dataset and use a 1-sample t-test to compare your estaimated sample mean to the population mean. Can you say that they are different? 

In [154]:
Highway_mpg=df['highway-mpg'].sample(n=30, random_state=30)

print(Highway_mpg.mean())


32.46666666666667


Import stats from scipy



In [155]:
from scipy.stats import stats

Null Hypothesesis the average highway mpg is  equal to 35mpg

Atlernative theory is Average Highway mpg is NOT equal to 35mpg

95% confidence level 

In [158]:
stats.ttest_1samp(Highway_mpg, 35)

Ttest_1sampResult(statistic=-1.9960001864636792, pvalue=0.055400806249855286)

Based on a T-Test result of 1.45 & a Pvalue of 0.55 we fail to rejected the Null Hypothoses that average high MPG is 35

The salesman says the cars he sells typically have a fuel efficiency of about 35 miles per gallon on the highway. You want to verify his claim but can only test 10 cars. Using your sample of 10, test his claim and report your results.

In [159]:
Highway_mpg_2 = df['highway-mpg'].sample(n=10, random_state=30)
print(Highway_mpg_2.mean())
stats.ttest_1samp(Highway_mpg_2, 35)

32.8


Ttest_1sampResult(statistic=-0.9570244044334747, pvalue=0.36355472283248624)

Based on a T-test of -0.95 and P value of 0.36 we fail to reject the Hypothesis 

Imagine that you now have the capacity to test 100 cars on the lot. Using the same random state of `30`, take a sample of 100 cars. Run a t-test to verify the salesman's claim with your new larger sample. Do you reach the same conclusion as you did with the sample size of 10? 

In [160]:
### YOUR WORK HERE
Highway_mpg3 = df['highway-mpg'].sample(n=100, random_state =30)
print(Highway_mpg3.mean())
stats.ttest_1samp(Highway_mpg3, 35)

30.64


Ttest_1sampResult(statistic=-6.5185242467032305, pvalue=3.003137417468754e-09)

Why might these two t-tests using the same dataset lead to different conclusions about the salesman's claim?

Based on a T test of 6.5 and a P-value of 0.000000003 we have rejected the hypothesis of 35mpg

## Practice 2-Sample T-tests

Two Sample t-tests determine whether or not two sample means are statistically different from each other. 

This portion of your assignment requires you to determine which issues have "statistically significant" differences between political parties in this [1980s congressional voting data](https://archive.ics.uci.edu/ml/datasets/Congressional+Voting+Records). 


### 1) Load the data

The data consists of 435 instances (one for each congressperson), a class (democrat or republican), and 16 binary attributes (yes or no for voting for or against certain issues). Be aware - there are missing values!

- Read the dataset in from UCI, you'll need to provide a list of column headers
- Encode "yes" votes as 1 and "no" votes as 0. (You can use `df.replace()` to swap out these values)
- Use dataframe filtering to split the dataframe into two new dataframes based on party. Hold all republicans in one dataframe and all democrats in the other. These will be our two different "samples."



In [161]:
### YOUR WORK HERE
from google.colab import files
uploaded = files.upload()


Saving house-votes-84.data to house-votes-84 (2).data


In [162]:
column_headers2 = ['class-name','handicapped-infants','water-project-cost-sharing','adoption-of-the-budget-resolution','physician-fee-freeze',
                   'el-salvador-aid','religious-groups-in-schools','anti-satellite-test-ban','aid-to-nicaraguan-contras',
                   'mx-missile','immigration','synfuels-corporation-cutback',
                   'education-spending','superfund','crime','duty-free-exports',
                   'export-administration']
votes = pd.read_csv('house-votes-84.data', names=column_headers2) 
votes.head(10)                  

class-name handicapped-infants  ... duty-free-exports export-administration
0  republican                   n  ...                 n                     y
1  republican                   n  ...                 n                     ?
2    democrat                   ?  ...                 n                     n
3    democrat                   n  ...                 n                     y
4    democrat                   y  ...                 y                     y
5    democrat                   n  ...                 y                     y
6    democrat                   n  ...                 y                     y
7  republican                   n  ...                 ?                     y
8  republican                   n  ...                 n                     y
9    democrat                   y  ...                 ?                     ?

[10 rows x 17 columns]

In [164]:
votes2 = votes.replace({'y': 1 , 'n':0, '?':np.NaN})
votes2.head(10)

class-name  handicapped-infants  ...  duty-free-exports  export-administration
0  republican                  0.0  ...                0.0                    1.0
1  republican                  0.0  ...                0.0                    NaN
2    democrat                  NaN  ...                0.0                    0.0
3    democrat                  0.0  ...                0.0                    1.0
4    democrat                  1.0  ...                1.0                    1.0
5    democrat                  0.0  ...                1.0                    1.0
6    democrat                  0.0  ...                1.0                    1.0
7  republican                  0.0  ...                NaN                    1.0
8  republican                  0.0  ...                0.0                    1.0
9    democrat                  1.0  ...                NaN                    NaN

[10 rows x 17 columns]

In [165]:
condition = votes2['class-name'] == 'republican'
rep = votes2[condition]
rep


class-name  handicapped-infants  ...  duty-free-exports  export-administration
0    republican                  0.0  ...                0.0                    1.0
1    republican                  0.0  ...                0.0                    NaN
7    republican                  0.0  ...                NaN                    1.0
8    republican                  0.0  ...                0.0                    1.0
10   republican                  0.0  ...                0.0                    0.0
..          ...                  ...  ...                ...                    ...
427  republican                  0.0  ...                0.0                    1.0
430  republican                  0.0  ...                0.0                    1.0
432  republican                  0.0  ...                0.0                    1.0
433  republican                  0.0  ...                0.0                    1.0
434  republican                  0.0  ...                NaN                    0.0

[168 rows x 17 columns]

In [166]:
condition2 = votes2['class-name']=='democrat'
dem = votes2[condition2]
dem

class-name  handicapped-infants  ...  duty-free-exports  export-administration
2     democrat                  NaN  ...                0.0                    0.0
3     democrat                  0.0  ...                0.0                    1.0
4     democrat                  1.0  ...                1.0                    1.0
5     democrat                  0.0  ...                1.0                    1.0
6     democrat                  0.0  ...                1.0                    1.0
..         ...                  ...  ...                ...                    ...
425   democrat                  0.0  ...                1.0                    NaN
426   democrat                  1.0  ...                1.0                    1.0
428   democrat                  NaN  ...                1.0                    1.0
429   democrat                  1.0  ...                1.0                    1.0
431   democrat                  0.0  ...                0.0                    1.0

[267 rows x 17 columns]

### 2) Perform two sample T-tests on different issues and report the results.

- Find an issue that democrats support more than republicans with p < 0.01 (significant at the 99% level).
- Find an issue that republicans support more than democrats with p < 0.01 (significant at the 99% level).
- Find an issue where the difference between republicans and democrats has p > 0.1 (Not significant at the 90% level - i.e. there may not be much of a difference the two sample means)

Please for each test that you run state your null and alternative hypothesis and then write a conclusion reflecting on the null and alternative hypothesis.

Remember, that two-sample t-tests will only tell us if the two groups are *different* from one another. We'll have to look at their sample means directly or use the sign on the t-statistic to know which of the two sample means is larger. 

In [167]:
import scipy.stats as sp

In [168]:
print(rep)
rep['physician-fee-freeze'].value_counts


     class-name  handicapped-infants  ...  duty-free-exports  export-administration
0    republican                  0.0  ...                0.0                    1.0
1    republican                  0.0  ...                0.0                    NaN
7    republican                  0.0  ...                NaN                    1.0
8    republican                  0.0  ...                0.0                    1.0
10   republican                  0.0  ...                0.0                    0.0
..          ...                  ...  ...                ...                    ...
427  republican                  0.0  ...                0.0                    1.0
430  republican                  0.0  ...                0.0                    1.0
432  republican                  0.0  ...                0.0                    1.0
433  republican                  0.0  ...                0.0                    1.0
434  republican                  0.0  ...                NaN                

<bound method IndexOpsMixin.value_counts of 0      1.0
1      1.0
7      1.0
8      1.0
10     1.0
      ... 
427    1.0
430    1.0
432    1.0
433    1.0
434    1.0
Name: physician-fee-freeze, Length: 168, dtype: float64>

In [169]:
print(dem.head(10))
dem['physician-fee-freeze'].value_counts


   class-name  handicapped-infants  ...  duty-free-exports  export-administration
2    democrat                  NaN  ...                0.0                    0.0
3    democrat                  0.0  ...                0.0                    1.0
4    democrat                  1.0  ...                1.0                    1.0
5    democrat                  0.0  ...                1.0                    1.0
6    democrat                  0.0  ...                1.0                    1.0
9    democrat                  1.0  ...                NaN                    NaN
12   democrat                  0.0  ...                NaN                    NaN
13   democrat                  1.0  ...                1.0                    NaN
16   democrat                  1.0  ...                0.0                    1.0
17   democrat                  1.0  ...                1.0                    1.0

[10 rows x 17 columns]


<bound method IndexOpsMixin.value_counts of 2      NaN
3      0.0
4      0.0
5      0.0
6      1.0
      ... 
425    0.0
426    0.0
428    0.0
429    0.0
431    0.0
Name: physician-fee-freeze, Length: 267, dtype: float64>

In [170]:
stats.ttest_ind(dem['physician-fee-freeze'], rep['physician-fee-freeze'], nan_policy='omit')

Ttest_indResult(statistic=-49.36708157301406, pvalue=1.994262314074344e-177)

Based on a t-test of -49.36 and pvalue of 1.99 e-177 we can say that Rep strongly support 'physician-fee'freeze'

In [171]:
dem.head(20)

class-name  handicapped-infants  ...  duty-free-exports  export-administration
2    democrat                  NaN  ...                0.0                    0.0
3    democrat                  0.0  ...                0.0                    1.0
4    democrat                  1.0  ...                1.0                    1.0
5    democrat                  0.0  ...                1.0                    1.0
6    democrat                  0.0  ...                1.0                    1.0
9    democrat                  1.0  ...                NaN                    NaN
12   democrat                  0.0  ...                NaN                    NaN
13   democrat                  1.0  ...                1.0                    NaN
16   democrat                  1.0  ...                0.0                    1.0
17   democrat                  1.0  ...                1.0                    1.0
19   democrat                  1.0  ...                1.0                    1.0
20   democrat                  1.0  ...                1.0                    1.0
21   democrat                  1.0  ...                1.0                    1.0
22   democrat                  1.0  ...                1.0                    1.0
23   democrat                  1.0  ...                1.0                    1.0
24   democrat                  1.0  ...                1.0                    NaN
25   democrat                  1.0  ...                1.0                    1.0
26   democrat                  1.0  ...                1.0                    1.0
27   democrat                  1.0  ...                1.0                    1.0
29   democrat                  1.0  ...                1.0                    1.0

[20 rows x 17 columns]

In [172]:
dem['adoption-of-the-budget-resolution'].value_counts()


1.0    231
0.0     29
Name: adoption-of-the-budget-resolution, dtype: int64

In [173]:
stats.ttest_ind(dem['adoption-of-the-budget-resolution'],rep['adoption-of-the-budget-resolution'], nan_policy='omit')

Ttest_indResult(statistic=23.21277691701378, pvalue=2.0703402795404463e-77)

Based on a Ttest value of 23 and p value of  2.7 to e-77 we can conclude that democrats value this bill more than rep

In [202]:
print(dem['water-project-cost-sharing'].value_counts())
print(rep['water-project-cost-sharing'].value_counts())
stats.ttest_ind(dem['water-project-cost-sharing'], rep['water-project-cost-sharing'], nan_policy='omit')

1.0    120
0.0    119
Name: water-project-cost-sharing, dtype: int64
1.0    75
0.0    73
Name: water-project-cost-sharing, dtype: int64


Ttest_indResult(statistic=-0.08896538137868286, pvalue=0.9291556823993485)

with a T-test of -0.88 and a Pvalue of .929 we cannot reject the null hypothesis that they support that rep and dem suppor the bill equally

## Stretch Goals:

### 1) Use functions and some iterator (for loop, .apply(), list comprehension, etc.) to perform two sample t-tests on every issue in the dataset in an automated fashion.

In [ ]:
### YOUR WORK HERE



### 2) Work on Performing a T-test without using Scipy in order to get "under the hood" and learn more thoroughly about this topic.
### Start with a 1-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://blog.minitab.com/hs-fs/hubfs/Imported_Blog_Media/701f9c0efa98a38fb397f3c3ec459b66.png?width=247&height=172&name=701f9c0efa98a38fb397f3c3ec459b66.png) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

### Be sure to check your work using Scipy!





In [ ]:
### YOUR WORK HERE

### 3) Then try a 2-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://lh3.googleusercontent.com/proxy/rJJ5ZOL9ZDvKOOeBihXoZDgfk7uv1YsRzSQ1Tc10RX-r2HrRpRLVqlE9CWX23csYQXcTniFwlBg3H-qR8MKJPBGnjwndqlhDX3JxoDE5Yg) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

 ### Be sure to check your work using Scipy!

In [ ]:
### YOUR WORK HERE